File: Data Structures & Algorithms/minimum-window-with-characters/submission-2.java
class Solution {
        public static String minWindow(String s, String t) {
        if(s.length() < t.length()) return "";

        int[] arr = new int[126];

        for(char c : t.toCharArray()){
            arr[c]++;
        }
        
        int left = 0; int right =0;
        int minLength = Integer.MAX_VALUE;
        int required = t.length();
        int start  = 0;
        
        while(right < s.length()){
            char rChar = s.charAt(right);
            
            if(arr[rChar] > 0){
                required--;
            }
            arr[rChar]--;
            right++;
            
            while(required == 0){
                if(right - left < minLength){
                    minLength = right - left;
                    start = left;
                }
                char lChar = s.charAt(left);
                arr[lChar]++;
                if(arr[lChar] > 0){
                    required++;
                }
                left++;
            }
        }
        return minLength == Integer.MAX_VALUE ? "" : s.substring(start, start + minLength);
    }


}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.kt
class Solution {
    fun characterReplacement(s: String, k: Int): Int {
    var left = 0
    var maxFreq = 0
    var maxLength = 0
    var arr = IntArray(26)

    for(right in s.indices){
        arr[s[right]- 'A']++
        maxFreq = maxOf(maxFreq, arr[s[right] - 'A'])

        while((right - left +1) - maxFreq > k){
            arr[s[left]- 'A']--
            left++
        }
        maxLength = maxOf(maxLength, right- left +1)
    }
    return maxLength

}
}
